### ISSUE TITLE
i18n (internationalization)

### ISSUE THREAD

**Issue by vizax** [MegaMek/megamek#103]
Starting this topic for discussion about adjusting the project(s) to be compatible for i18n/internationalization of strings.

Thoughts/Comments welcome.

**Comment by arlith** [comment]
PR #75 talks about this a bit, particularly giving directions on how to enable warnings for string literal usages.

What kinds of guidelines were you looking for?  Basically, anything that is text that could be displayed to a user should be internationalized.  Anything that never gets displayed to the user should be marked as NLS.  The messages are pulled from properties files in the i18n directory.

**Comment by djmbm** [comment]
Reminder, that only applies to MegaMek.

MekHQ took a different approach, and embeds the strings in resource files instead.

Not sure what MML does.

And to make it even worse, MML and MHQ both use MM's as well, since they use MM code which uses them.

**Comment by Akjosch** [comment]
Compared to the amount of work in externalising those strings (ugh ... just got done with *one* class in MHQ), changing where the properties files are located and how they are loaded is more or less just a quick find-and-replace job.

**Comment by vizax** [comment]
I don't want to go overboard with this, but at least this way anyone who helps will have some guidelines and will make sure all future updates follow the same path. Think of it as making sure everyone is on the same page.

Guidelines questions:
- Naming conventions for variables?
  - Examples: `class.variable` or just `variable` or `class.subclass.variable`. Some existing properties files use like the above, but others use numbers.
- Name conventions `.Properties` files (or maybe `.Properties` file per what breakdown)?
  - Examples: `messages.properties` or `class_messages.properties`. Some existing `messages.properties` files are a single file, others are multiple files for different classes.
- Communicating who is working on what to remove the possibility of double work.
  - Do we create an Issue (or just a PR) for each class we complete? Do we (you devs) allow multiple message files to submitted at once?

just a few things that come to mind to help with organization.

**Comment by Akjosch** [comment]
My preference would be ...

* multiple files (at least one per package, with a "common" one for common texts)
* `variable.text` for straight texts, `variable.format` for things you put into `String.format(...)`, and similar suffixed denoting what the variable is *for*
* keeping the variable names as short as possible while remaining legible and unique - so *no* numbers

**Comment by pheonixstorm** [comment]
Unless the format is all over the place, best to look up how MM handles all the variables as is for now. If not then this project will double in size as all current work will need to be reworked to match the new format as well. Not sure if the current dev team wants it that way but that is how I would if it was my project. If you wont see it through to the end don't bother starting it :stuck_out_tongue_winking_eye: :poop:

**Comment by vizax** [comment]
@pheonixstorm the format is not the same across files right now.

And yes, you are absolutely correct that the devs need to decide if this is something they want to actually tackle. I've already volunteered my assistance, should they decide this needs to be accomplished. But, this isn't a one or two person job, if we ever want it completed within any reasonable time frame. Hence the discussion :grin: If we can set guidelines then anyone can pick it up and knock out a section, thus allowing the more experienced coders more time to do the cool stuff. :+1: :+1:

**Comment by arlith** [comment]
I'm not sure how many files we really want to have. Right now, there are 4 messages.properties files (which can then have any number of messages_<language>.properties):  `megamek/client/`, `megamek/client/bot`, `megamek/common`, and `megamek/common/options`.  There's also `megamek/common/reports-messages.properties` which uses numbers to match reports.

The problem I see with having many files is that it becomes more difficult to know just which messages.properties file (for language variant) that you need to modify.

Another thing that we need in this vein is a utility that will tell us which messages have an English translation, but are missing a translation in another language.

**Comment by nderwin** [comment]
@arlith That's a common problem - reconciling missing translations.  I've used Checkstyle for this, and probably some others I can't recall at the moment.  Attached is a report generated from it (sans style).  
[checkstyle.html.zip](https://github.com/MegaMek/megamek/files/190678/checkstyle.html.zip)

It notes that there are 2858 missing translations across all of the .properties files.  And two source files with tabs in them... hrm...

Something else to consider, as I ran into this with a previous project, is that the .properties files are expected to be encoded as ISO 8859-1, so any funky characters (like Cyrillic for Russian) need to go through the native2ascii tool to be used in the file.  Or, you dig into the guts of Properties and make it open the file stream with UTF-8 (we did the latter).

**Comment by vizax** [comment]
Will this be sufficient for doing replacements?  [old Eclipse Tutorial for i18n](http://www.eclipse.org/articles/Article-Internationalization/how2I18n.html) (basics still apply, by the look of it)  I've tested it out and it seems pretty simple. 

**It will auto-create or change:**
- The `messages.properties` for each package (`megamek`, `megamek.client`, `megamek.client.bot`, etc.), along with code replacements in whatever substitution pattern you declare.
- A Messages.java for each package
- Whatever changes you do to your code
![image](https://cloud.githubusercontent.com/assets/1894549/14069674/9e6f0f04-f462-11e5-9562-2326c1c1ecb9.png)
![image](https://cloud.githubusercontent.com/assets/1894549/14069678/afa8218e-f462-11e5-8c16-d16dad05f9fe.png)
![image](https://cloud.githubusercontent.com/assets/1894549/14069705/f98ca5b8-f462-11e5-96a0-55a781fe2f2d.png)

**Comment by pheonixstorm** [comment]
So an idea for messages.properties. To keep the file from exploding into something difficult to manage we could break it ino topic_messages_language.properties So all planetary conditions would be one file. Pilot options and quirks could go into another. Unit type and unit related items could be a 3rd etc. Thoughts?

**Comment by vizax** [comment]
FWIW, I like the way Eclipse does it. One `message.properties` per package, and everything is identified within by class. IMO, makes it easy to find where you need to go. I'll go with whatever the devs decide and then I'll work with that.

`megamek.client` package (i18n folder) snapshot:
![image](https://cloud.githubusercontent.com/assets/1894549/14130521/1713c93a-f5f9-11e5-9829-e51db0ef56c7.png)

*Not that I'd want to do it, but with the Eclipse system (which uses numbers) 1 or 2 people could knock this out in a day or 2.  I like, and think it is a good idea, to have something more descriptive. But if you are going for speed, then it could work.*

**Comment by Dylan-M** [comment]
So, here's the problem:
We have a lot of strings that are used in multiple places (packages). This is why the structure exists the way it is now.

I do like the idea of breaking them up a little more based on _what_ they are, but even then you'll need a "common" file for messages that are very generic and may be used in multiple places and ways.

**Comment by Dylan-M** [comment]
Additionally, I labeled this as a question since it isn't actually a bug per se.

**Comment by arlith** [comment]
I think that we currently use something similar to the Eclipse system, no?  If a string has multiple references, I don't think it's too big of a burden to just  duplicate it (who's to say it may not change in the future?)

**Comment by vizax** [comment]
Unless there is a different preference, I'll branch my fork in a bit and use the eclipse method and get started.

**Comment by vizax** [comment]
Started. Visit: [https://github.com/vizax/megamek/tree/i18n](https://github.com/vizax/megamek/tree/i18n) to see progress or let me know if i need to do things differently.

**Comment by arlith** [comment]
It appears you may have tabs on, instead of spaces (there's an option to replace tabs with spaces).  I say this because looking at your diffs, it looks like the files you update have changes on every single line.  This will make merging those changes really difficult.

**Comment by vizax** [comment]
I did have tabs on. I'll see about fixing it. Though, I saw that 1 commit had a full file changed (I was going to ask before I submitted a PR, but hadn't gotten around to it), I didn't see any others.

**Comment by arlith** [comment]
In the pull request page, there will be a tab for a list of commits and a tab for the final diff.  If you fix the tabs/spaces issue, I think that the final diff should be cleaner.

**Comment by vizax** [comment]
well, I'm not sure this worked right. 
I did the replace tabs with space (which is set to 4 spaces), and formatted, and some of it came out different. (i.e. - MegaMek.java and RandomNameGenerator.java both still show full file changes)
check my steps based on : [http://stackoverflow.com/a/22213976](http://stackoverflow.com/a/22213976)

at this point I guess i should rebase everything and start over (at least i can keep the properties files and just reuse names)  :hankey: :disappointed:

**Comment by arlith** [comment]
Those directions look like they come from an older Eclipse.  That option still exists, but I think it's just for the text editor.  If you want the Java code formatter to use spaces, you also need to go to Preferences, and then Java > Code Style > Formatter and then edit your preferred profile.  On the formatter profile edit panel, there's an "Indentation" tab (it's the default tab) there's a "Tab policy" combo box, where you can select "spaces only".  This is what you have to change to get the code formatter to use spaces.

**Comment by elementx54** [comment]
I remember from when I implemented the Russian translation bulk and then proceeded to fix why having a dedicated server in one languages with a clients in another produced weird text for weather reports, _that_ the rabbit hole goes a lot deeper than you'd think at first glance. Now this was a long time ago, but I found that translation of text sometimes happens on the server side and sometimes on the client side. It would be ideal to have all translations happen on the client side or the server side, but not both. I personally would choose client side. Just saying that fixing translations would mean messing with TONS of code in all kinds of places. And then you'd have to think about the other "Mek friend" programs and how they do it.

**Now** for the real reason I am resurrecting this issue: Would a unit test, or whatever test structure we have right now, benefit from a simple, non-critical, warning producing test that basically compares the English `*.properties` file values to see if they are covered in the other language `*.properties` file values? Then a simple warning message of alerting "_#_ of pairs not found in _language_" for each language is shown at compile time. It wouldn't halt anything but it would show something is missing and get annoying. :stuck_out_tongue: 

I'll *try* to get caught up again on stuff with i18n.
